cmd/queuebench/main.go

```diff
@@ -0,0 +1,159 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"log"
+	"time"
+
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	apmqueue "github.com/elastic/apm-queue"
+	"github.com/elastic/apm-queue/kafka"
+)
+
+const namespace = "queuebench"
+
+var totalproduced = 0
+var totalconsumed = 0
+
+func main() {
+	// NOTE: intercept any panic and print a nice terminate gracefully
+	// This allows using log.Panic methods in main; those function
+	// trigger deferred functions whereas log.Fatal don't.
+	defer func() {
+		if r := recover(); r != nil {
+			log.Fatal(r)
+		}
+	}()
+
+	cfg := config{}
+
+	cfg.Parse()
+
+	fmt.Printf("%+v\n", cfg)
+
+	log.Println("prep logger")
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		log.Fatalf("cannot create zap logger: %s", err)
+	}
+
+	log.Println("prep MeterProvider")
+	rdr := sdkmetric.NewManualReader()
+	mp := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(rdr),
+	)
+
+	ctx := context.Background()
+
+	bench := bench{
+		Brokers:        []string{cfg.broker},
+		Logger:         logger,
+		Partitions:     cfg.partitions,
+		TopicNamespace: namespace,
+		Topics: []apmqueue.Topic{
+			// TODO: replace with a unique identifier for a single run
+			apmqueue.Topic(fmt.Sprintf("run-%d", time.Now().Unix())),
+		},
+
+		mp: mp,
+		tp: trace.NewNoopTracerProvider(),
+	}
+
+	log.Println("running benchmark setup")
+	if err = bench.Setup(ctx); err != nil {
+		log.Panicf("benchmark setup failed: %s", err)
+	}
+	defer func() {
+		log.Println("running benchmark teardown")
+		if err := bench.Teardown(ctx); err != nil {
+			log.Panicf("benchmark teardown failed: %s", err)
+		}
+	}()
+
+	start := time.Now()
+	log.Println("==> running benchmark")
+
+	log.Println("start consumer")
+	go func() {
+		if err := bench.c.Run(ctx); err != nil {
+			log.Panicf("consumer run ended with an error: %s", err)
+		}
+	}()
+
+	log.Println("start producing")
+	produce(ctx, bench.p, bench.Topics[0], cfg.eventSize)
+
+	log.Println("stop producing")
+
+	for {
+		// TODO: add timeout
+		if totalconsumed < totalproduced {
+			time.Sleep(100 * time.Millisecond)
+		} else {
+			break
+		}
+	}
```

Comment:
Another option here could be to:
 - when the producer is done, tell the consumer how many messages were produced
 - have the consumer signal when it has consumed that many messages (e.g. by closing a channel)

Then this code can just block until that is signalled, or until the context is cancelled. This should minimise observer effect.

Reply:
I worked on this implementation today but I have some doubts about it being the right solution considering the trade off with simplicity.

I refactored the implementation to use a ticker and react to a timeout (default `1m`) and I wondered if the a delay of at maximum 100ms is worth choosing the channel based implementation. We could also reduce the ticker time to a lesser value, the assumption on which I chose 100ms is I considered it an effective trade off: benchmarks will run for seconds if not minutes so a skew of < 99ms would not impact the overall result significantly.

My preferred approach here would be to leverage otel metrics emitted by producer and consumer (replacing the globals in use now), but this would not fit the channel based design.

Reply:
I was thinking more about CPU consumption that polling adds. I'm fine with using the OTel metrics for this if the overhead of gathering the metrics every 100ms isn't significant. The current implementation isn't concurrent-safe, as when there are multiple partitions there may be multiple consumer goroutines. Relying on either a channel approach or OTel metrics should resolve that.

Reply:
> I was thinking more about CPU consumption that polling adds. I'm fine with using the OTel metrics for this if the overhead of gathering the metrics every 100ms isn't significant. 

I will test and report back. One question about the CPU consumption: I was under the assumption that we aim to establish a baseline, not to make a realistic estimation of throughput. The implication to me is that as long as conditions are consistent across benchmark we shouldn't worry too much on ensuring the benchmark code performs as its best.
What's your thoughts on this?

> The current implementation isn't concurrent-safe, as when there are multiple partitions there may be multiple consumer goroutines. Relying on either a channel approach or OTel metrics should resolve that.

Indeed, I'll investigate the CPU consumption for using otel metrics and decide which approach to choose base don it.

Reply:
I profiled the code and my conclusion is that the impact is not significant, but I'd like to review the results tomorrow with @axw

Reply:
> One question about the CPU consumption: I was under the assumption that we aim to establish a baseline, not to make a realistic estimation of throughput. The implication to me is that as long as conditions are consistent across benchmark we shouldn't worry too much on ensuring the benchmark code performs as its best.
What's your thoughts on this?

It's not so much about comparing one run to another, but about identifying the maximum throughput so we compare against users of apm-queue. It can be approximate, but the closer to the true maximum the better.

Reply:
We reviewed the profiling data and confirmed that using otel metrics does not impact the benchmark.